File: 프로그래머스/3/118669. 등산코스 정하기/등산코스 정하기.cpp
#include <string>
#include <vector>
#include <queue>
#include <algorithm>

#define MAX 50000+1
#define INF 10000000+1

using namespace std;

vector<pair<int, int>> adj[MAX];
bool isSummit[MAX];
int intensity[MAX];

vector<int> solution(int n, vector<vector<int>> paths, vector<int> gates, vector<int> summits) {
    vector<int> answer;
    fill(intensity, intensity+n+2, INF);
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
    
    for(auto path: paths) {
        adj[path[0]].push_back({path[2], path[1]});
        adj[path[1]].push_back({path[2], path[0]});
    }
    for(auto summit: summits) {
        isSummit[summit] = true;
    }
    for(auto gate: gates) {
        intensity[gate] = 0;
        pq.push({0, gate});
    }
    
    while(!pq.empty()) {
        int curIntensity = pq.top().first;
        int curIdx = pq.top().second;
        pq.pop();
        
        if(intensity[curIdx] != curIntensity) continue;
        
        for(auto nxt: adj[curIdx]) {
            int nxtWeight = nxt.first;
            int nxtIdx = nxt.second;
            
            if(intensity[nxtIdx] <= max(curIntensity, nxtWeight)) continue;
            intensity[nxtIdx] = max(curIntensity, nxtWeight);
            
            if(isSummit[nxtIdx]) continue;
            pq.push({intensity[nxtIdx], nxtIdx});
        }
    }
    
    int minIntensity = INF, minIdx;
    sort(summits.begin(), summits.end());
    for(auto summit: summits) {
        if(intensity[summit] < minIntensity) {
            minIntensity = intensity[summit];
            minIdx = summit;
        }
    }
    
    answer.push_back(minIdx);
    answer.push_back(minIntensity);
    
    return answer;
}

File: 프로그래머스/3/118669. 등산코스 정하기/등산코스 정하기.py
import heapq
from collections import defaultdict

MAX = 10000001

def solution(n, paths, gates, summits):
    answer = [-1, MAX]
        
    is_summit = [False] * (n + 1)
    for summit in summits:
        is_summit[summit] = True

    adjs = defaultdict(list)
    for i, j, w in paths:
        adjs[i].append((j, w))
        adjs[j].append((i, w))
        
    intensity = [MAX] * (n + 1)
        
    pq = list()
    for gate in gates:
        heapq.heappush(pq, (0, gate))
        intensity[gate] = 0
        
    while pq:
        cur_intensity, cur_node = heapq.heappop(pq)
        
        if intensity[cur_node] < cur_intensity:
            continue
            
        for nxt_node, nxt_intensity in adjs[cur_node]:
            nxt_intensity = max(cur_intensity, nxt_intensity)
            
            if intensity[nxt_node] <= nxt_intensity:
                continue
                
            intensity[nxt_node] = nxt_intensity
            
            if is_summit[nxt_node]:
                continue
                
            heapq.heappush(pq, (nxt_intensity, nxt_node))
    
    summits.sort()
    for summit in summits:
        if intensity[summit] < answer[1]:
            answer = [summit, intensity[summit]]
    
    return answer